### PR TITLE
Add Half TensorType support

### PR DESCRIPTION
**Pull request by ehsanmok**
This PR adds `f16` support using [half crate](https://github.com/starkat99/half-rs) with minimal dependency.

#196

### CODE REVIEW

**Comment by adamcrume** [comment]
There's a lot of unrelated formatting changes that make it difficult to see the real delta.  Please remove the formatting changes from this PR.

**Comment by ehsanmok** [comment]
No worries, that's why I asked it [here](#196) earlier whether to include the formatting changes. I'd be easier to merge #203 first, then I rebase after.

**Comment by ehsanmok** [comment]
@adamcrume Please check when you have time.

File: src/lib.rs [hunk 1]
```diff
@@ -1530,6 +1530,21 @@ mod tests {
...
+        let data: Vec<f16> = vec![-1.0f32, 0.0, 1.0, 2.5]
+            .into_iter()
+            .map(|x| f16::from_f32(x))
+            .collect();
+        let tensor = <Tensor<f16>>::new(&[2, 2]).with_values(&data);
+        assert!(tensor.is_ok());
+        let tensor = tensor.unwrap();
+        assert_eq!(tensor.len(), 4);
+        for i in 0..data.len() {
+            assert_eq!(tensor[i], data[i]);
```
Comment (adamcrume) [inline src/lib.rs]: You can make this more succinct by replacing

```
assert_eq!(tensor.len(), 4);
for i in 0..data.len() {
    assert_eq!(tensor[i], data[i]);
}
```

with

```
assert_eq!(&tensor[..], &data[..]);
```

Comment (ehsanmok) [reply src/lib.rs]: Sure :) done

**Comment by adamcrume** [comment]
Thanks!

**Pull request merged**